src/Graphics/OpenGLContext/opengl_BufferedDrawer.cpp

```diff
@@ -0,0 +1,218 @@
+#include <Graphics/Context.h>
+#include <Config.h>
+#include "GLFunctions.h"
+#include "opengl_Attributes.h"
+#include "opengl_CachedFunctions.h"
+#include "opengl_BufferedDrawer.h"
+
+using namespace graphics;
+using namespace opengl;
+
+BufferedDrawer::BufferedDrawer(const GLInfo & _glinfo, CachedVertexAttribArray * _cachedAttribArray, CachedBindBuffer * _bindBuffer)
+: m_glInfo(_glinfo)
+, m_cachedAttribArray(_cachedAttribArray)
+, m_bindBuffer(_bindBuffer)
+{
+	m_cachedAttribArray->enableVertexAttribArray(triangleAttrib::position, false);
+	m_cachedAttribArray->enableVertexAttribArray(triangleAttrib::color, false);
+	m_cachedAttribArray->enableVertexAttribArray(triangleAttrib::texcoord, false);
+	m_cachedAttribArray->enableVertexAttribArray(triangleAttrib::numlights, false);
+	m_cachedAttribArray->enableVertexAttribArray(triangleAttrib::modify, false);
+
+	m_cachedAttribArray->enableVertexAttribArray(rectAttrib::position, false);
+	m_cachedAttribArray->enableVertexAttribArray(rectAttrib::color, false);
+	m_cachedAttribArray->enableVertexAttribArray(rectAttrib::texcoord0, false);
+	m_cachedAttribArray->enableVertexAttribArray(rectAttrib::texcoord1, false);
+
+	int fake;
+	m_attribsData.fill(&fake);
+	m_bufMaxSize = 4194304;
```

Comment:
It probably is overkill, it's the same reason I did ```m_size * 32``` in the PBO persistent mapping code.

When the buffer is full, it resets the offset back to 0 and starts over at the beginning of the buffer. There is nothing preventing you from overwriting something in the buffer that hasn't been used yet, which would result in corruption/crash. Basically I made the buffer large enough that I can be pretty sure by the time I've "made my way around the buffer", the contents aren't neede anymore.

The safer way to do it would be to use a sync object to make sure it is safe to overwrite a portion of the buffer, but I've never ran into any problems doing it this way (it's been like this in GLupeN64 for a while now and I've never had a single issue). I figured using 12MB of memory (3 buffers) is pretty small and then I don't have to have the overhead of the sync objects.